### ISSUE TITLE
Configure sources to skip proxy

### ISSUE THREAD

**Issue by mrdima** [chocolatey/choco#262]
While setting up a demo I ran into the undesired result of the chocolatey client (0.9.9.5) taking automatically the proxy configuration of Internet Explorer (in LAN) settings. This is actually fine when going to the Internet, but in this case our (simple) chocolatey server is installed on a machine within the same network as I’m running the chocolatey client on…or on an internal network anyway.
Even if I specify “Bypass proxy server for local addresses” or in the Advanced/Exceptions “Do not use proxy server for addresses beginning with:” add 10.1.1 it doesn’t work, chocolatey is always trying to use the proxy server to connect to the internal address.
Without IE proxy settings:
clist -source http://10.1.1.46/chocolatey
Chocolatey v0.9.9.5
Whatever 1.93
BvSshServer 6.07
2 packages found.

Eg, working fine…
Then with proxy settings:
clist -source http://10.1.1.46/chocolatey
Chocolatey v0.9.9.5
The remote server returned an error: (503) Server Unavailable.

Not working because the proxy server can’t reach the internal address and we don’t want that either…chocolatey should use a direct connection.
So in a lot of cases where you use the public repo this auto-proxy use is perfect, but unfortunately not in our case. May I request to have a config option to overrule the proxy behavior and be able to at least set it to “do not use proxy, use direct connection”? Or is there some other method already? It’s really  a breaking thing for us because if someone configures a proxy server in IE then our repository cannot be reached anymore and no more installations/upgrades etc are possible.

**Comment by ferventcoder** [comment]
Thanks for reporting this.

**Comment by JonCubed** [comment]
Happy to test as well when a fix is available

**Comment by da9l** [comment]
Unless this already has been resolved I'd like to vote +1 on this. When we want to use chocolatey in our build pipeline this creates problems for us.  

We often have projects that has dependencies both to packages on external feeds and to internal packages in internal http based feeds.

Meanwhile does anyone have a suggestion for a workaround for this problem?

**Comment by ferventcoder** [comment]
As a workaround, just make sure those sources are included in your proxy sources to skip. Do you have an explicit proxy enabled or is it just using the system one? The system one is configurable on what to skip.

**Comment by ferventcoder** [comment]
It should work, but that really depends on whether NuGet.Core handles it appropriately. Try with the latest beta of Chocolatey to be sure.

**Comment by sfhardman** [comment]
I'm using the following as a workaround on Chocolatey v0.9.10-beta1-281-gfd2eea1:

- Explictly set chocolatey proxy (`choco config set proxy http://proxy.domain:port`)
- Add a hosts file entry for your internal NuGet server so that it can be accessed without a dot in the host name (see [WebProxy.IsBypassed docs](https://msdn.microsoft.com/en-us/library/system.net.webproxy.isbypassed(v=vs.110).aspx))
- Add the chocolatey source using the name registered in the hosts file, without any domain name

**Comment by maxgaulhofer** [comment]
Why does choco get proxy settings from IE?
I would expect that if i configure an proxy in chocolatey.config like this
```xml
  <config>
    <add key="proxy" value="" description="Explicit proxy location." />
    <add key="proxyUser" value="" description="Optional proxy user." />
    <add key="proxyPassword" value="" description="Optional proxy password. Encrypted." />
  </config>
```
That it does not use a proxy at all ("") = no proxy

Why does choco import the settings from IE by default and doesnt it look at the settings configured with netsh winhttp show proxy? Is this a design decision?

**Comment by ferventcoder** [comment]
Probably a carryover - it implicitly uses the proxy. It shouldn't use it if it is bypassed but some of that code goes into NuGet.Core (which may not make the best decision about bypassed proxies).

If you can send me a log with `--debug --verbose` we can make a determination where this is occurring.

**Comment by ferventcoder** [comment]
@maxgaulhofer I think the issue you want is #497 - that is where we are adding `choco config set proxyBypassList`

**Comment by ferventcoder** [comment]
@mrdima @da9l @sfhardman In 0.10.4 this will be: 

`choco source add -n --bypassProxy`

This will still need to be added to the `chocolateysource` puppet provider. cc @jpogran / @glennsarti / @Iristyle /  @chocolatey/chocolatey-puppet

**Comment by ferventcoder** [comment]
For Puppet - filed https://tickets.puppetlabs.com/browse/MODULES-4418 to handle adding the `bypass_proxy` to `chocolateysource`.

**Comment by ferventcoder** [comment]
> Why does choco import the settings from IE by default and doesn't it look at the settings configured with netsh winhttp show proxy? Is this a design decision?

@maxgaulhofer Design decision. It's a fallback to catch proxy configuration set for the system.

**Comment by ferventcoder** [comment]
Still working on this - the Choco/NuGet.Core does a connection to the domain prior to hitting the source and that is causing some issues.

**Comment by ferventcoder** [comment]
This is now fixed for 0.10.4
